lmu2png/src/main.cpp

```diff
@@ -14,79 +14,242 @@
 #undef main
-    // =========================================================================
-    int main(int argc, char** argv)
-    {
-        const char * usage = "Usage: lmu2png map.lmu chipset.png output.png\n";
-
-        if(argc < 4 || argc > 5)
-        {
-            std::cout<<usage;
-            exit(EXIT_FAILURE);
-        }
-
-        const char* map_path = argv[1];
-        const char* chipset_path = argv[2];
-        const char* output_path = argv[3];
-
-        std::unique_ptr<RPG::Map> map = LMU_Reader::Load(map_path, "");
-        if (map.get() == NULL)
-        {
-            std::cerr<<LcfReader::GetError()<<std::endl;
-            exit(EXIT_FAILURE);
-        }
-
-        SDL_Surface* chipset = IMG_Load(chipset_path);
-        if (chipset == NULL)
-        {
-            std::cerr<<IMG_GetError()<<std::endl;
-            exit(EXIT_FAILURE);
-        }
-
-        if (IMG_isBMP(SDL_RWFromFile(chipset_path, "rb")))
-        {
-            // Set as color key the first color in the palette
-            SDL_Color ckey = chipset->format->palette->colors[0];
-            SDL_SetColorKey(chipset, SDL_TRUE, SDL_MapRGB(chipset->format, ckey.r, ckey.g, ckey.b));
-        }
-
-        SDL_Surface* output = SDL_CreateRGBSurface(0, map->width * 16, map->height * 16, 8, 0, 0, 0, 0);
-        output->format->palette = chipset->format->palette;
-        stChipset gen;
-        gen.GenerateFromSurface(chipset);
-
-        for (int y = 0; y < map->height; ++y)
-            for (int x = 0; x < map->width; ++x)
-            {
-                gen.RenderTile(output, x*16, y*16, map->lower_layer[x+y*map->width], 0);
-                gen.RenderTile(output, x*16, y*16, map->upper_layer[x+y*map->width], 0);
-            }
-
-        std::vector<RPG::Event>::iterator ev;
-        for (ev = map->events.begin(); ev != map->events.end(); ++ev)
-        {
-            RPG::EventPage evp = ev->pages[0];
-            if (evp.character_name.empty())
-                gen.RenderTile(output, (ev->x)*16, (ev->y)*16, 0x2710 + evp.character_index, 0);
-        }
-
-        if(IMG_SavePNG(output, output_path) < 0)
-        {
-	    std::cerr<<IMG_GetError()<<std::endl;
-	    exit(EXIT_FAILURE);
-        }
-
-        exit(EXIT_SUCCESS);
+
+const std::string usage = R"(Usage: lmu2png map.lmu [options]
+Options:
+	-d database.ldb
+	-c chipset.png
+	-e encoding
+	-o output.png
+	--no-background
+	--no-lowertiles
+	--no-uppertiles
+	--no-events
+)";
+
+std::string GetFileDirectory (const std::string& file) {
+	size_t found = file.find_last_of("/\\");
+	return found == std::string::npos ? "./" : file.substr(0,found + 1);
+}
+
+bool Exists(const std::string& filename) {
+    std::ifstream infile(filename.c_str());
+    return infile.good();
+}
+
+std::string path;
+
+std::string FindResource(const std::string& folder, const std::string& base_name) {
+	static const std::string rtp2k(getenv("RPG2K_RTP_PATH"));
+	static const std::string rtp2k3(getenv("RPG2K3_RTP_PATH"));
```

Comment:
These two lines throw on Linux (logic_error) and segfault (0 pointer) on Windows. Needs an extra-check if getenv returns nullptr.